libcob/profiling.c

```diff
@@ -0,0 +1,445 @@
+	curr_info = called_runtimes[current_idx];
+
+	if (info == curr_info && is_in_same_section (info, proc_idx, curr_proc)) {
+		while (current_idx >= 0) {
+			curr_info = called_runtimes[current_idx];
+			curr_proc = called_procedures[current_idx];
+			if (is_section (curr_info, curr_proc)) {
+				break;
+			}
+			cob_prof_exit_section (curr_info, curr_proc);
+		}
+
+		if (is_section (info, proc_idx)) { /* GO TO current section */
+			return;
+		} else {
+			cob_prof_enter_paragraph (info, proc_idx);
+		}
+	} else {
+		/* If not in the same section, then we exit every
+		 * procedure before entering the targeted one. */
+		while (current_idx >= 0) {
+			curr_info = called_runtimes[current_idx];
+			curr_proc = called_procedures[current_idx];
+			if (is_section (curr_info, curr_proc)) {
+				cob_prof_exit_section (curr_info, curr_proc);
+			} else {
+				cob_prof_exit_paragraph (curr_info, curr_proc);
+			}
+		}
+
+		if (is_section (info, proc_idx)) {
+			cob_prof_enter_section (info, proc_idx);
+		} else {
+			cob_prof_enter_paragraph (info, proc_idx);
+		}
+	}
+}
+
+
+static void
+print_location (FILE *file, struct cobprof_info *info, int proc_num, const char* kind) {
+	char buf[COB_NORMAL_BUFF];
+	const char *name = info->procedures_names[proc_num];
+	int len = strlen (name);
+	int i;
+
+	strcpy (buf, name);
+	for (i=0; i<len; i++){
+		if (buf[i] == '|') buf[i] = ',';
+	}
+	fprintf (file, "%s,%s,%s,", info->program_id, buf, kind);
+}
+
+static void
+print_monotonic_time (FILE *file, cob_ns_time t) {
+
+	cob_ns_time nanoseconds = t ;
+	unsigned int hours = 0;
+	unsigned int minutes = 0;
+	unsigned int seconds = 0;
+	unsigned int milliseconds = 0;
+
+	if (nanoseconds >= 100000) {
+		milliseconds = t / 100000;
+		nanoseconds %= 100000;
+	}
+	if (milliseconds >= 1000) {
+		seconds = milliseconds / 1000;
+		milliseconds %= 1000;
+	}
+	if (seconds >= 60) {
+		minutes = seconds / 60;
+		seconds %= 60;
+	}
+	if (minutes >= 60) {
+		hours = minutes / 60;
+		minutes %= 60;
+	}
+
+	fprintf (file, "%Ld,", t);
+
+	if (hours > 0) {
+		fprintf (file, "%u h %u m %u s %u ms %Lu ns", hours, minutes, seconds,
+			 milliseconds, nanoseconds);
+	} else if (minutes > 0) {
+		fprintf (file, "%u m %u s %u ms %Lu ns", minutes, seconds, milliseconds,
+			 nanoseconds);
+	} else if (seconds > 0) {
+		fprintf (file, "%u s %u ms %Lu ns", seconds, milliseconds, nanoseconds);
+	} else if (milliseconds > 0) {
+		fprintf (file, "%u ms %Lu ns", milliseconds, nanoseconds);
+	} else {
+		fprintf (file, "%Lu ns", nanoseconds);
+	}
+}
+
+void
+cob_prof_end ()
+{
+	FILE *file;
+	char prof_file_buf[COB_NORMAL_BUFF];
+	const char* prof_filename = NULL;
+
+	if (!cobsetptr || !is_active || !prof_program_id) return;
+
+	while (current_idx >= 0) {
+		cob_prof_exit_section (called_runtimes[current_idx], called_procedures[current_idx]);
+	}
+
+	prof_filename = cobsetptr->cob_prof_filename;
+	if (!prof_filename){
+
+		/* Do not use cob_sys_getpid() in case of fork() */
```

Comment:
That's a very interesting point. We should possibly adjust the function instead.